compiler/rustc_middle/src/mir/syntax.rs

```diff
@@ -180,6 +180,53 @@ pub enum BorrowKind {
     Mut { kind: MutBorrowKind },
 }
 
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, TyEncodable, TyDecodable)]
+#[derive(Hash, HashStable)]
+pub enum RawPtrKind {
+    Mut,
+    Not,
+    /// Represents a raw pointer for a place to access its mutability.
+    /// The data behind this point will not be accessed, just its metadata.
```

Comment:
```suggestion
    /// Creates a raw pointer to a place that will only be used to access its metadata,
    /// not the data behind the pointer. Note that this limitation is *not* enforced
    /// by the validator.
    /// The borrow checker allows overlap of these raw pointers with references. This
    /// is sound even if the pointer is "misused" since any such use is anyway unsafe.
    /// In terms of the operational semantics (i.e., Miri(, this is equivalent to `Mut`, but
    /// will never incur a retag.
```
I assume "access its mutability" is a typo.

Reply:
> this is equivalent to `Mut`, but will never incur a retag.

Why is this equivalent to `Mut`? We're never mutating through this pointer.

Reply:
non-`Mut` strips the mutability permission from the provenance, which this does not. Therefore it is closer to `&raw mut` than to `&raw const`.

Reply:
Frankly, I think it should go the other way -- fake borrows should probably strip *all* permissions from the provenance. You shouldn't be reading or writing data through this point after all?

Reply:
We could decide that but then the Miri implementation should actually do this.

Reply:
> We could decide that but then the Miri implementation should actually do this.

Why? Miri doesn't guarantee to check all UB anyways. And remember that the previous implementation was literally using span hacks here along with `&raw const` pointers.

Anyways, if you think that this change would be simple to implement, please feel free to point out how that can be done and I can apply it, otherwise I'm leaning towards leaving this as-is.

Reply:
Every case of Miri missing UB is a critical bug in Miri. So no it's not acceptable to document one thing here and implement something else in Miri.

Reply:
I think it is simple to implement these fake borrows as being like `&raw mut`, and I don't see any significant downsides to this approach, whereas the alternatives proposed so far (including the current state of the PR) all do have significant downsides.

Reply:
> Anyways, if you think that this change would be simple to implement, please feel free to point out how that can be done and I can apply it, otherwise I'm leaning towards leaving this as-is.

Note that the implementation you give it in Miri makes the resulting pointer have all permissions of its parent. Unless you plan to change that, this should be documented.

Reply:
I mean, I'm happy to (and probably would be more comfortable) with making miri understand that these pointers have no permissions, I just don't know (1.) whether that's feasible or (2.) whether that's easy. It would be consistent with the checks we implement in borrowck, after all.

Reply:
I thought borrowck ignored these fake raw borrows entirely?

I'd prefer for Miri to just preserve the provenance (and therefore all permissions) here; IMO it's not worth adding the complexity required to strip provenance. As already mentioned, I am working towards making *all* `&raw` just preserve provenance, and I don't want this one to be a permanent exception.

Reply:
K, I guess I'll just update the comment.